### PR TITLE
ascanrules: correct tag search in HtmlContextAnalyser

### PR DESCRIPTION
**Pull request by kcberg**
Was seeing this stacktrace in the most recent update to the XSS rule that was causing the rule to crash and not find the reflected injection.

```
java.lang.StringIndexOutOfBoundsException: begin 3, end 1, length 28
	at java.lang.String.checkBoundsBeginEnd(String.java:4604) ~[?:?]
	at java.lang.String.substring(String.java:2707) ~[?:?]
	at org.zaproxy.zap.extension.ascanrules.httputils.HtmlContextAnalyser.parseTag(HtmlContextAnalyser.java:172) ~[?:?]
	at org.zaproxy.zap.extension.ascanrules.httputils.HtmlContextAnalyser.getHtmlContexts(HtmlContextAnalyser.java:308) ~[?:?]
	at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRule.performAttack(CrossSiteScriptingScanRule.java:240) ~[?:?]
	at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRule.performAttack(CrossSiteScriptingScanRule.java:177) ~[?:?]
	at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRule.performAttack(CrossSiteScriptingScanRule.java:154) ~[?:?]
	at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRule.performTagAttack(CrossSiteScriptingScanRule.java:364) ~[?:?]
	at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRule.scan(CrossSiteScriptingScanRule.java:812) ~[?:?]
	at org.parosproxy.paros.core.scanner.AbstractAppParamPlugin.scan(AbstractAppParamPlugin.java:201) ~[zap-2.12.0-h0.0.41-SNAPSHOT.jar:2.12.0-h0.0.41-SNAPSHOT]
	at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRule.scan(CrossSiteScriptingScanRule.java:145) ~[?:?]
	at org.parosproxy.paros.core.scanner.AbstractAppParamPlugin.scan(AbstractAppParamPlugin.java:126) ~[zap-2.12.0-h0.0.41-SNAPSHOT.jar:2.12.0-h0.0.41-SNAPSHOT]
	at org.parosproxy.paros.core.scanner.AbstractAppParamPlugin.scan(AbstractAppParamPlugin.java:87) ~[zap-2.12.0-h0.0.41-SNAPSHOT.jar:2.12.0-h0.0.41-SNAPSHOT]
	at org.parosproxy.paros.core.scanner.AbstractPlugin.run(AbstractPlugin.java:330) ~[zap-2.12.0-h0.0.41-SNAPSHOT.jar:2.12.0-h0.0.41-SNAPSHOT]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

Signed-off-by: KC Berg <kc.berg@stackhawk.com>

### CODE REVIEW

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyser.java [hunk 1]
```diff
@@ -169,6 +169,9 @@ public static Map<String, Map<String, String>> parseTag(String input) {
             if (tagEnd == -1) {
                 break;
             }
+            if (tagStart + 1 > tagEnd) {
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyser.java:174]: Like 169 thru 174

```
if (tagEnd == -1 || tagStart + 1 > tagEnd) {
    break;
}
```

**Review by kingthorin — APPROVED** [review]
Thanks

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyser.java [hunk 2]
```diff
@@ -166,7 +166,7 @@ public static Map<String, Map<String, String>> parseTag(String input) {
                 break;
             }
             tagEnd = input.indexOf('>');
-            if (tagEnd == -1) {
+            if (tagEnd == -1 || tagStart + 1 > tagEnd) {
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyser.java:169]:
```suggestion
            tagEnd = input.indexOf('>', tagStart);
            if (tagEnd == -1) {

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyserUnitTest.java [hunk 3]
```diff
@@ -164,4 +164,14 @@ void shouldNotBreakParsingTagsInInvalidHtml() {
         tagMap = HtmlContextAnalyser.parseTag(tag);
         assertThat(tagMap.size(), is(equalTo(3)));
     }
+
+    @Test
+    void shouldNotRaiseStringIndexOutOfBoundsException() {
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyserUnitTest.java:169]: `shouldParseWithUnescapedTagEndChar`

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyserUnitTest.java [hunk 4]
```diff
@@ -164,4 +164,14 @@ void shouldNotBreakParsingTagsInInvalidHtml() {
...
+        String tag = "<html> <body> <span>></span> </body> </html>";
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/httputils/HtmlContextAnalyserUnitTest.java:170]:
```suggestion
        String tag = "<html> <body> <span>></span> <a></a> </body> </html>";
```
To verify the parsing does not end earlier than needed, browsers keep parsing.

**Comment by thc202** [comment]
The changelog should be updated since this was breaking the scan rule.

File: addOns/ascanrules/CHANGELOG.md [hunk 5]
```diff
@@ -13,6 +13,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Fixed
 - The Remote File Inclusion scan rule no longer follows redirects before checking the response for content indicating a vulnerability (Issue 5887).
 - False positive where Cross Site Scripting payloads are safely rendered in a textarea tag.
+- Unescaped tag end causing Cross Site Scripting rule to crash.
```
Comment (kingthorin) [inline addOns/ascanrules/CHANGELOG.md:16]: Maybe "throw an exception" instead of "crash".

**Review by kingthorin — COMMENTED** [review]
Looks good to me, I'd just suggest a minor change log tweak.

**Review by kingthorin — APPROVED** [review]

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**